### ISSUE TITLE
Why calling (*Connection) Close / Cancel doesn't close corresponding amqp.Delivery channel?

### ISSUE THREAD

**Issue by inliquid** [streadway/amqp#405]
In this example application hangs after `Ctrl-C` because `consumer` never breaks out of a `amqp.Delivery` channel `range` loop:

 ```go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"

	"github.com/streadway/amqp"
)

var (
	shutdown = make(chan struct{})
)

func main() {
	go signalListener()
	go consumer()
	go publisher()

	fmt.Scanln()
}

func consumer() {
	cnt := 0
	defer func() { log.Printf("Consumer exited, got %d messages", cnt) }()

	conn, ch, q := getQueue()
	defer conn.Close()
	defer ch.Close()

	msgs, err := ch.Consume(
		q.Name,
		"someConsumer",
		true,
		false,
		false,
		false,
		nil)

	errPanic(err, "Failed to register a consumer")

	for msg := range msgs {
		cnt++
		_ = msg
		//log.Println("Received message:", string(msg.Body))
	}
}

func publisher() {
	cnt := 0
	defer func() { log.Printf("Publisher exited, sent %d messages", cnt) }()

	conn, ch, q := getQueue()
	defer conn.Close()
	//defer func() { log.Println(ch.Close()) }() // <-- WHY amqp.Delivery not closed after this
	defer func() { log.Println(ch.Cancel("someConsumer", false)) }() // <-- WHY amqp.Delivery not closed after this

	msg := amqp.Publishing{
		ContentType: "text/plain",
		Body:        []byte("Hello RabbitMQ"),
	}

	for {
		select {
		case <-shutdown:
			log.Printf("Stopping publisher and closing '%s' queue\n", q.Name)
			return
		default:
			ch.Publish("", q.Name, false, false, msg)
			cnt++
		}
	}
}

func getQueue() (*amqp.Connection, *amqp.Channel, *amqp.Queue) {
	conn, err := amqp.Dial("amqp://guest@localhost:5672")
	errPanic(err, "Failed to connect to RabbitMQ")
	ch, err := conn.Channel()
	errPanic(err, "Failed to open a channel")
	q, err := ch.QueueDeclare("hello",
		false,
		false,
		false,
		false,
		nil)
	errPanic(err, "Failed to declare a queue")

	return conn, ch, &q
}

func errPanic(err error, msg string) {
	if err != nil {
		log.Panicf("%s: %v\n", msg, err)
	}
}

func signalListener() {
	defer close(shutdown)

	s := make(chan os.Signal, 1)
	signal.Notify(s, os.Interrupt)

	log.Println("Signal caught:", <-s)
}
```

**Comment by michaelklishin** [comment]
I'll leave it to @streadway to provide some historical API design context but both options makes certain sense. The question comes down to whether the library should manage [Go] channels for you. Other libraries do not always have the same abstraction but the Erlang one does and consumer processes are not stopped by the client when the consumer is cancelled. Connection closure usually results in a a termination of a whole tree of processes but again, the consumer one specifically is not managed by the library.

**Comment by inliquid** [comment]
@michaelklishin ok, in that case, what is the preferred or idiomatic way of stopping consumers with this lib?

**Comment by michaelklishin** [comment]
Cancel the consumer, closes its [protocol] channel (and, if needed, connection) and then dispose of any resources it might have been using (such as Go channels).

**Comment by inliquid** [comment]
@michaelklishin let me rephrase, what would be a correct way for consumer to catch an event of cancelled producer, as done in example app?

**Comment by streadway** [comment]
In this example, the consumer and producer have different connections. The producer's AMQP channel doesn't have a consumer called "someConsumer".

Call `Channel.Cancel()` on the consumer's Channel to let the server drain the deliveries to the consumer's AMQP channel and then let the library drain the deliveries to the receiver of the `chan amqp.Delivery` returned from `Channel.Consume()`.

Since there can be more than one consumer per AMQP channel, you'd also need to specify the same consumer tag for both the `Channel.Consume()` and `Channel.Cancel()` calls.

You can order the last producer's `Channel.Publish` before the consumer's `Channel.Cancel()` in a few different ways. I prefer to create refined interfaces so the logic can be tested independent from the implementation.  I prefer to separate connection handling from pub/sub logic. I also prefer to be able to test pub/sub logic with different implementations of the connection. Doing it that way would look something like:

```
func main() {
  // this is newConsumer()
  sub := &consumer{tag: "someConsumer"}
  sub.conn, sub.ch, sub.q = getQueue()

  // this is newLinkedProducer(link interface { Cancel() error })
  pub := &linkedProducer{link: sub}
  pub.conn, pub.ch, pub.q = getQueue()

  go publisher(pub)
  go consumer(sub)

  // signal handler and other stuff ...
}

type Publisher interface {
  Close() error
  Publish(msg amqp.Publishing) error
}

type linkedPublisher struct {
  conn *amqp.Connection
  ch *amqp.Channel
  q *amqp.Queue

  link interface { Cancel() error }
}

// Close closes the publisher connection before the cancelling the consumer.
func (pub *linkedPublisher) Close() error {
  if err := pub.conn.Close(); err != nil {
    return err
  }
  return pub.link.Cancel()
}

func (pub *linkedPublisher) Publish(msg amqp.Publishing) error {
  return pub.ch.Publish("", pub.q.Name, false, false, msg)
}

func producer(pub Publisher) {
  defer pub.Close()

   msg := amqp.Publishing{
   	ContentType: "text/plain",
   	Body:        []byte("Hello RabbitMQ"),
   }
   pub.Publish("", q.Name, false, false, msg)
}

type Consumer interface {
  Close() error
  Cancel() error
  Consume() (<-chan amqp.Delivery, error)
}

type consumer struct {
  conn *amqp.Connection
  ch *amqp.Channel
  q *amqp.Queue
  tag string
}

func (c *consumer) Close() error {
  return c.conn.Close()
}

func (c *consumer) Consume() (<-chan amqp.Delivery, error) {
  return c.ch.Consume(
   	c.q.Name,
   	c.tag,
   	true,
   	false,
   	false,
   	false,
   	nil)
}

func (c *consumer) Cancel() error {
  return c.ch.Cancel(c.tag, false)
}

func consumer(sub Consumer) {
  defer sub.Close()
  ch, err := sub.Consume()
  for msg := <-ch {
    ...
  }
}
```

An alternative approach, closer to this example, is to use another package level channel for coordination.

```
var cancel = make(chan struct{})

func consumer() {
  ...
  conn, ch, q := getQueue()
  ...
  defer func() {
    <-cancel
    ch.Cancel("someConsumer", false)
  }()
  ...
}

func producer() {
  defer close(cancel)
  ...
}
```

**Comment by inliquid** [comment]
Thank you very much for a detailed answer. May I just ask you briefly for a part I didn't quite get?
You're saying 
>The producer's AMQP channel doesn't have a consumer called "someConsumer".

But isn't this call supposed to establish one?
```go
   msgs, err := ch.Consume(
   	q.Name,
   	"someConsumer", // <-- THIS
   	true,
   	false,
   	false,
   	false,
   	nil)
```
> Call Channel.Cancel() on the consumer's Channel

I also tried both `ch.Cancel(<...>)` and `ch.Close()`
```go
   defer conn.Close()
   //defer func() { log.Println(ch.Close()) }() // <-- WHY amqp.Delivery not closed after this
   defer func() { log.Println(ch.Cancel("someConsumer", false)) }() // <-- WHY amqp.Delivery not closed after this
```

**Comment by streadway** [comment]
Consumers are scoped to a Channel, a Channel is scoped to a Connection and a Connection is scoped to a TCP session.

Exchanges and Queues are global to broker, so it can be surprising that "Consumer" is local (and protected) to a Channel. One channel cannot cancel a consumer in a different channel.

You just need to call `Cancel()` on the same channel you called `Consume()`.

**Comment by inliquid** [comment]
@streadway thank you!
